pennylane/math/decomposition.py

```diff
@@ -436,60 +509,50 @@ def givens_decomposition(U):
-    M, N = math.shape(unitary)
+    shape = math.shape(unitary)
+
+    if len(shape) != 2 or shape[0] != shape[1]:
+        raise ValueError(f"The unitary matrix should be of shape NxN, got {shape}")
 
-    if M != N:
-        raise ValueError(f"The unitary matrix should be of shape NxN, got {unitary.shape}")
+    N = shape[0]
 
     left_givens, right_givens = [], []
     for i in range(1, N):
         if i % 2:
-            for j in range(0, i):
+            for j in range(i):
                 indices = [i - j - 1, i - j]
-                grot_mat = _givens_matrix(*unitary[N - j - 1, indices].T, left=True)
+                grot_mat = _givens_matrix(*unitary[N - j - 1, indices].T, real_valued=is_real)
                 unitary = _set_unitary_matrix(
                     unitary, (Ellipsis, indices), unitary[:, indices] @ grot_mat.T, like=interface
                 )
                 right_givens.append((math.conj(grot_mat), indices))
         else:
-            for j in range(1, i + 1):
-                indices = [N + j - i - 2, N + j - i - 1]
-                grot_mat = _givens_matrix(*unitary[indices, j - 1], left=False)
+            for j in range(i):
+                indices = [N + j - i - 1, N + j - i]
+                grot_mat = _givens_matrix(*unitary[indices, j], left=False, real_valued=is_real)
                 unitary = _set_unitary_matrix(
                     unitary, (indices, Ellipsis), grot_mat @ unitary[indices, :], like=interface
                 )
                 left_givens.append((grot_mat, indices))
+    if is_real:
+        return _absorb_phases_so(left_givens, right_givens, N, unitary, interface)
 
-    nleft_givens = []
-    for grot_mat, (i, j) in reversed(left_givens):
-        sphase_mat = math.diag(math.diag(unitary)[math.array([i, j])])
-        decomp_mat = math.conj(grot_mat).T @ sphase_mat
-        givens_mat = _givens_matrix(*decomp_mat[1, :].T)
-        nphase_mat = decomp_mat @ givens_mat.T
-
-        # check for T_{m,n}^{-1} x D = D x T.
-        if not math.is_abstract(decomp_mat) and not math.allclose(
-            nphase_mat @ math.conj(givens_mat), decomp_mat
-        ):  # pragma: no cover
-            raise ValueError("Failed to shift phase transposition.")
-
-        for diag_idx, diag_val in zip([(i, i), (j, j)], math.diag(nphase_mat)):
-            unitary = _set_unitary_matrix(unitary, diag_idx, diag_val, like=interface)
-        nleft_givens.append((math.conj(givens_mat), (i, j)))
-
-    phases, ordered_rotations = math.diag(unitary), []
-    for grot_mat, (i, j) in list(reversed(nleft_givens)) + list(reversed(right_givens)):
-        if not math.is_abstract(grot_mat) and not math.all(
-            math.isreal(grot_mat[0, 1]) and math.isreal(grot_mat[1, 1])
-        ):  # pragma: no cover
-            raise ValueError(f"Incorrect Givens Rotation encountered, {grot_mat}")
-        ordered_rotations.append((grot_mat, (i, j)))
-
-    return phases, ordered_rotations
```

Comment:
This was moved into a function `_commute_phases_u` above. I removed both loops of checks, because the only way they would trigger an error would be by changing the definition of `_givens_matrix`. Checking that this does not happen is the job of the test suite, not of runtime code.